sys/linux/dev_usbmon.txt

```diff
@@ -0,0 +1,88 @@
+# Copyright 2017 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+# https://elixir.bootlin.com/linux/latest/source/Documentation/usb/usbmon.txt
+# https://elixir.bootlin.com/linux/latest/source/drivers/usb/mon/mon_bin.c
+
+include <linux/fcntl.h>
+include <linux/ioctl.h>
+
+define MON_IOC_MAGIC	0x92
+
+define SETUP_LEN	8
+define PAGE_SIZE	4096
+define CHUNK_SIZE	PAGE_SIZE
+
+define BUFF_MAX	(((1200*1024)+CHUNK_SIZE-1) & ~(CHUNK_SIZE-1))
+define BUFF_MIN	(((8*1024)+CHUNK_SIZE-1) & ~(CHUNK_SIZE-1))
+
+define MON_IOCQ_URB_LEN	_IO(MON_IOC_MAGIC, 1)
+define MON_IOCG_STATS	_IOR(MON_IOC_MAGIC, 3, struct {u32 queued; u32 dropped;})
+define MON_IOCT_RING_SIZE	_IO(MON_IOC_MAGIC, 4)
+define MON_IOCQ_RING_SIZE	_IO(MON_IOC_MAGIC, 5)
+define MON_IOCX_GET	_IOW(MON_IOC_MAGIC, 6, struct {void *hdr; void *data; size_t alloc;})
+define MON_IOCX_MFETCH	_IOWR(MON_IOC_MAGIC, 7, struct {u32 *offvec; u32 nfetch; u32 nflush;})
+define MON_IOCH_MFLUSH	_IO(MON_IOC_MAGIC, 8)
+define MON_IOCX_GETX	_IOW(MON_IOC_MAGIC, 10,  struct {void *hdr; void *data; size_t alloc;})
+
+resource fd_usbmon[fd]
+
+openat$usbmon(fd const[AT_FDCWD], file ptr[in, string["/dev/usbmon0"]], flags flags[open_flags], mode const[0]) fd_usbmon
```

Comment:
@xairy we need some USB expertise here. Please take a look too.
I also wonder if this will make interesting interactions with your USB fuzzing?
Also, usbmon0 is created always, but I see that usbmon1+ can also be created. When? Will we get 1 per test proc on usb-fuzzer? How do we need to open usbmon1+?

Reply:
`usbmon` is a USB monitoring interface that allows to sniff USB packets. I'm not sure if there's much interaction between USB drivers fuzzing and this. `usbmon0` exposes USB packets from all USB buses present on the system, `usbmon1` - only from bus #1, etc. So to have more `usbmonN` devices, you need to have more USB buses. One way to do this is to enable `CONFIG_USB_DUMMY_HCD` (that's what we do for USB fuzzing), which is going to create one more virtual USB bus. Another way to pass some flags to QEMU, but that won't work on syzbot.

Reply:
I have 33 /dev/usbmon's and CONFIG_USB_DUMMY_HCD is not enabled.
When I enabled CONFIG_USB_DUMMY_HCD I have 34 of them.
Perhaps this one creates 32 usbmon's:
```
CONFIG_USBIP_VHCI_NR_HCS=16
 ```
I wonder which one is CONFIG_USB_DUMMY_HCD? Is it /dev/usbmob1 or /dev/usbmon33?
I think we should at least add:
```
syz_open_dev$usbmon(dev ptr[in, string["/dev/usbmon#"]], id proc[1, 2], flags flags[open_flags]) fd_usbmon
```

Reply:
Please add
```
syz_open_dev$usbmon(dev ptr[in, string["/dev/usbmon#"]], id proc[1, 2], flags flags[open_flags]) fd_usbmon
```
in a follow up PR. With syzbot config we have 34 /dev/usbmon* files. Such syscall will allow syzkaller to open these additional devices.